File: guests/haskell/sliders.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DeriveAnyClass #-}

-- The sliders scene, Haskell port — guests/rust/sliders.rs,
-- tools/scenes/sliders.steps, docs/slider-plan.md.

import Data.IORef (atomicModifyIORef', newIORef)
import Data.List (isSuffixOf)
import GHC.Generics (Generic)
import Text.Printf (printf)

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

data Track = Track {name :: Text, level :: Double}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)


-- The harness's own slider spelling (crates/kaya/src/harness.rs).
-- printf is String by nature (Text.Printf has no Text instance), so the
-- one conversion sits here rather than at every call.
spelled :: Double -> Text
spelled v = T.pack (dropDot (dropZeros (printf "%.6f" v)))
  where
    dropZeros s = if "0" `isSuffixOf` s then dropZeros (init s) else s
    dropDot s = if "." `isSuffixOf` s then init s else s

keyWord :: Key -> Text
keyWord = keyText

main :: IO ()
main = kayaMain $ \app -> do
  commitsRef <- newIORef (0 :: Int)

  (commitText, rowText, master, levelNode) <- buildTx app $ do
    levelText <- signalText "value: 50"
    commitText <- signalText "commits: 0"
    volumeText <- signalText "volume: 0.5"
    rowText <- signalText "row: none"
    pos <- signalDouble 50.0
    tracks <- collectionOf @Track

    let onLevel v =
          submitTx app $ writeSignal levelText ("value: " <> spelled v)
        onVolume v =
          submitTx app $ writeSignal volumeText ("volume: " <> spelled v)
        onReset =
          -- Must NOT come back as a value or a commit occurrence.
          submitTx app $ writeSignal pos (25.0 :: Double)

    master <-
      sliderBoundOn -- slider#0
        0.0
        100.0
        pos
        onLevel
        [Step 5.0, TickSpacing 25.0, A11yId "master", A11yLabel "Level"]

    (trackList, levelNode) <- forEach (recordHandle tracks) $ do
      -- Realized ahead of its row so the central registration has a handle.
      n <-
        withTplAttrs
          [TplStep 10.0, TplA11yId "level"]
          (slider 0.0 100.0 (field @"level" @Track))
      _ <- rowOf [label (field @"name" @Track), pure n]
      return n

    root <-
      column
        [ labelBound levelText, -- label#0
          labelBound commitText, -- label#1
          labelBound volumeText, -- label#2
          labelBound rowText, -- label#3
          pure master,
          sliderOn 0.0 1.0 0.5 onVolume [TickSpacing 0.25, A11yLabel "Volume"], -- slider#1
          buttonOn "reset" onReset, -- button#0
          pure trackList
        ]
    mount root

    insertRecord tracks "a" (Track "a" 70.0)
    insertRecord tracks "b" (Track "b" 20.0)
    return (commitText, rowText, master, levelNode)

  onValueCommitted app master $ \_ -> do
    n <- atomicModifyIORef' commitsRef (\c -> (c + 1, c + 1))
    submitTx app $ writeSignal commitText ("commits: " <> tshow n)

  onValueCommitted app levelNode $ \keys v -> case keys of
    (key : _) ->
      submitTx app $
        writeSignal rowText ("row " <> keyWord key <> ": " <> spelled v)
    [] -> error "kaya: onValueCommitted's key path is never empty"

File: guests/rust/sliders.rs
//! The sliders scene (tools/scenes/sliders.steps; docs/slider-plan.md): a
//! stepped slider with ticks bound to a signal so button#0 can write it
//! PROGRAMMATICALLY (the echo negative), a continuous one with ticks, and a
//! stamped stepped slider bound to a row's own field.

#[derive(kaya::KayaGen, Clone, Debug, PartialEq)]
struct Track {
    name: String,
    level: f64,
}

#[derive(Clone)]
enum Msg {
    Level(f64),
    Committed,
    Volume(f64),
    RowLevel(kaya::Path, f64),
    Reset,
}

fn key_word(path: &kaya::Path) -> String {
    match path.first() {
        Some(kaya::Value::Str(s)) => s.clone(),
        other => format!("{other:?}"),
    }
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let mut commits = 0u32;
    let (level_text, commit_text, volume_text, row_text, pos, level_node) = ctx.apply(|tx| {
        let level_text = tx.signal("value: 50");
        let commit_text = tx.signal("commits: 0");
        let volume_text = tx.signal("volume: 0.5");
        let row_text = tx.signal("row: none");
        let pos = tx.signal(50.0);
        let tracks = tx.collection::<Track>();
        let mut level_node = kaya::TemplateNodeId(0);
        let root = tx
            .column(|tx| {
                tx.label(level_text); // label#0
                tx.label(commit_text); // label#1
                tx.label(volume_text); // label#2
                tx.label(row_text); // label#3
                let level = tx
                    .slider_bound(0.0, 100.0, pos) // slider#0
                    .step(5.0)
                    .tick_spacing(25.0)
                    .a11y_label("Level")
                    .id();
                tx.a11y_id(level, "master");
                msgs.on_value(level, Msg::Level);
                msgs.on_commit(level, |_| Msg::Committed);
                let volume = tx
                    .slider(0.0, 1.0, 0.5) // slider#1
                    .tick_spacing(0.25)
                    .a11y_label("Volume")
                    .id();
                msgs.on_value(volume, Msg::Volume);
                let reset = tx.button("reset").id(); // button#0
                msgs.on_click(reset, Msg::Reset);
                for mut row in tracks.rows(tx) {
                    row.label(Track::name());
                    let slider = row.slider(0.0, 100.0, Track::level());
                    row.step(slider, 10.0);
                    row.a11y_id(slider, "level");
                    level_node = slider;
                }
            })
            .id();
        tx.mount(root);
        tx.insert(&tracks, "a", Track { name: "a".into(), level: 70.0 });
        tx.insert(&tracks, "b", Track { name: "b".into(), level: 20.0 });
        (level_text, commit_text, volume_text, row_text, pos, level_node)
    });
    msgs.on_commit_node(level_node, Msg::RowLevel);

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Level(v) => ctx.apply(|tx| {
                tx.write(level_text, format!("value: {v}"));
            }),
            Msg::Committed => {
                commits += 1;
                ctx.apply(|tx| {
                    tx.write(commit_text, format!("commits: {commits}"));
                })
            }
            Msg::Volume(v) => ctx.apply(|tx| {
                tx.write(volume_text, format!("volume: {v}"));
            }),
            Msg::RowLevel(path, v) => ctx.apply(|tx| {
                tx.write(row_text, format!("row {}: {v}", key_word(&path)));
            }),
            Msg::Reset => ctx.apply(|tx| {
                // Must NOT come back as a value or a commit occurrence.
                tx.write(pos, 25.0);
            }),
        }
    }
}

fn main() {
    kaya::run(app)
}
